src/lib/openjp2/fbc_dec.c

```diff
@@ -0,0 +1,2534 @@
+  // CUP+SPP, and 3 for CUP+SPP+MRP
+  num_passes = cblk->numsegs>0 ? cblk->segs[0].real_num_passes : 0;
+  num_passes += cblk->numsegs>1 ? cblk->segs[1].real_num_passes : 0;
+  // OPJ_UINT32 lengths1 is the length of cleanup pass
+  lengths1 = num_passes > 0 ? cblk->segs[0].len : 0;
+  // OPJ_UINT32 lengths2 is the length of refinement passes (either SPP only or SPP+MRP)
+  lengths2 = num_passes > 1 ? cblk->segs[1].len : 0;
+  // OPJ_INT32 width is the decoded codeblock width 
+  width = cblk->x1 - cblk->x0;
+  // OPJ_INT32 height is the decoded codeblock height
+  height = cblk->y1 - cblk->y0;
+  // OPJ_INT32 stride is the decoded codeblock buffer stride 
+  stride = width;
+
+   /*  sigma1 and sigma2 contains significant (i.e., non-zero) pixel 
+    *  locations.  The buffers are used interchangeably, because we need
+    *  more than 4 rows of significance information at a given time.
+    *  Each 32 bits contain significance information for 4 rows of 8 
+    *  columns each.  If we denote 32 bits by 0xaaaaaaaa, the each "a" is
+    *  called a nibble and has significance information for 4 rows.
+    *  The least significant nibble has information for the first column,
+    *  and so on. The nibble's LSB is for the first row, and so on.
+    *  Since, at most, we can have 1024 columns in a quad, we need 128
+    *  entries; we added 1 for convenience when propagation of signifcance
+    *  goes outside the structure
+    *  To work in OpenJPEG these buffers has been expanded to 132.
+    */
+  // OPJ_UINT32 *pflags, *sigma1, *sigma2, *mbr1, *mbr2, *sip, sip_shift;
+  pflags = (OPJ_UINT32 *)t1->flags;
+  sigma1 = pflags;
+  sigma2 = sigma1 + 132;
+  // mbr arrangement is similar to sigma; mbr contains locations 
+  // that become significant during significance propagation pass
+  mbr1 = sigma2 + 132;
+  mbr2 = mbr1 + 132;
+  //a pointer to sigma
+  sip = sigma1;  //pointers to arrays to be used interchangeably
+  sip_shift = 0; //the amount of shift needed for sigma
+
+  if (num_passes > 1 && lengths2 == 0)
+  {
+    if (p_manager_mutex)
+      opj_mutex_lock(p_manager_mutex);
+    opj_event_msg(p_manager, EVT_WARNING, "A malformed codeblock that has "
+                  "more than one coding pass, but zero length for "
+                  "2nd and potential 3rd pass.\n");
+    if (p_manager_mutex)
+      opj_mutex_unlock(p_manager_mutex);
+    num_passes = 1;
+  }
+  if (num_passes > 3)
+  {
+    if (p_manager_mutex)
+      opj_mutex_lock(p_manager_mutex);
+    opj_event_msg(p_manager, EVT_WARNING, "We do not support more than 3 "
+                            "coding passes; This codeblocks has %d passes.\n",
+                            num_passes);
+    if (p_manager_mutex)
+      opj_mutex_unlock(p_manager_mutex);
+    return OPJ_FALSE;
+  }
+
+  if (cblk->numbps == 1 && num_passes > 1)
+    {
+      // We do not have enough precision to decode SgnProp nor MagRef passes.
```

Comment:
This is more of protection of the block decoder against calls with incorrect configuration.  In normal operation cblk->numbps cannot be smaller than 1, unless something is wrong.

Reply:
Some where in the email I received there was a message about 3 coding passes.  HT supports more than 3 coding passes, but only the last CUP, and if SPP and MRP exist after it, needs to be decoded; these passes are place holder passes.  I have not paid a lot of attention to the case of coding passes before that last CUP.  There are also some flags to signal this case.    

As far as I know, Kakadu can decode this case, but I am not very sure if it can generate this case.